pkg/core/sync.go

```diff
@@ -0,0 +1,210 @@
+package core
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/cloudquery/cloudquery/internal/logging"
+	"github.com/cloudquery/cloudquery/pkg/core/database"
+	"github.com/cloudquery/cloudquery/pkg/core/state"
+	"github.com/cloudquery/cloudquery/pkg/plugin"
+	"github.com/cloudquery/cloudquery/pkg/plugin/registry"
+	sdkdb "github.com/cloudquery/cq-provider-sdk/database"
+	"github.com/cloudquery/cq-provider-sdk/migration"
+	"github.com/cloudquery/cq-provider-sdk/migration/migrator"
+	"github.com/cloudquery/cq-provider-sdk/provider/diag"
+	"github.com/cloudquery/cq-provider-sdk/provider/execution"
+	"github.com/cloudquery/cq-provider-sdk/provider/schema"
+	"github.com/rs/zerolog/log"
+	"github.com/thoas/go-funk"
+)
+
+type SyncState int
+
+const (
+	Upgraded SyncState = iota + 1
+	Downgraded
+	NoChange
+)
+
+func (s SyncState) String() string {
+	switch s {
+	case Upgraded:
+		return "Upgraded"
+	case Downgraded:
+		return "Downgraded"
+	case NoChange:
+		return "NoChange"
+	}
+	return "Unknown"
+}
+
+type SyncOptions struct {
+	Provider registry.Provider
+	// DownloadLatest specifies whether the latest provider should be downloaded before sync is called.
+	DownloadLatest bool
+}
+
+type SyncResult struct {
+	State      SyncState
+	OldVersion string
+	NewVersion string
+}
+
+func Sync(ctx context.Context, storage database.Storage, pm *plugin.Manager, opts *SyncOptions) (*SyncResult, diag.Diagnostics) {
```

Comment:
important to note: every time sync is called and we have a new version we drop all data, i.e if I run a policy with latest version or any command that does sync (most commands do) we will drop data and policy won't work for example.

Reply:
the signature stuff takes care of that now